### ISSUE TITLE
`string.h` file not found for macOS

### ISSUE THREAD

**Issue by niyue** [corrosion-rs/corrosion#152]
# Setup
I use corrosion to build a C++ project with some Rust dependency, and the Rust dependency has a transitive dependency to this crate https://github.com/MaterializeInc/rust-krb5-src

# my environment
Operating system: macOS 12.3

```
❯ cc --version
Apple clang version 13.1.6 (clang-1316.0.21.2)
Target: x86_64-apple-darwin21.4.0
Thread model: posix
InstalledDir: /Library/Developer/CommandLineTools/usr/bin
```

# Issue
The above setup used to work for corrrosion f679545a63a8b214a415e086f910126ab66714fa, and I recently upgraded to a new version corrosion 0.1 release, and it starts to report the following error, and I tried a new commit (904fefbedff0a781ea6db861100f584326adb19c), which reports the same error as well. But if I rolled back to use f679545a63a8b214a415e086f910126ab66714fa, the error won't be reported. 

Could some one shed some light on it? Any help is appreciated. Thanks.

```
Compiling strip-ansi-escapes v0.1.1
error: failed to run custom build command for `krb5-src v0.3.2+1.19.2`
Caused by:
process didn't exit successfully: `/<redated>/x64-osx-dbg/cargo/build/debug/build/krb5-src-9aacd81c281a6b62/build-script-build` (exit status: 101)
--- stdout
checking for gcc... /Library/Developer/CommandLineTools/usr/bin/cc
checking whether the C compiler works... no
 --- stderr
configure: error: in `/<redated>/x64-osx-dbg/cargo/build/x86_64-apple-darwin/debug/build/krb5-src-6053a8bec92c223c/out/build':
configure: error: C compiler cannot create executables
See `config.log' for more details
thread 'main' panicked at 'configure failed: Custom { kind: Other, error: "command [\"./krb5/src/configure\", \"--prefix=/<redated>/x64-osx-dbg/cargo/build/x86_64-apple-darwin/debug/build/krb5-src-6053a8bec92c223c/out/install\", \"--enable-static\", \"--disable-shared\", \"--disable-nls\", \"CPPFLAGS=\", \"CFLAGS= -fPIC\"] exited with code 77" }', /Users/ss/.cargo/registry/src/github.com-1ecc6299db9ec823/krb5-src-0.3.2+1.19.2/build.rs:86:14
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
warning: build failed, waiting for other jobs to finish...
The following warnings were emitted during compilation:
warning: In file included from zstd/lib/common/entropy_common.c:18:
warning: In file included from zstd/lib/common/mem.h:24:
warning: zstd/lib/common/zstd_deps.h:29:10: fatal error: 'string.h' file not found
warning: #include <string.h>
warning: ^~~~~~~~~~
warning: 1 error generated.
warning: In file included from zstd/lib/common/fse_decompress.c:20:
warning: In file included from zstd/lib/common/bitstream.h:29:
warning: In file included from zstd/lib/common/mem.h:24:
warning: zstd/lib/common/zstd_deps.h:29:10: fatal error: 'string.h' file not found
warning: #include <string.h>
warning: ^~~~~~~~~~
warning: 1 error generated.
```

P.S. I did some search, and this error may be related with this similar issue (https://www.reddit.com/r/cmake/comments/clj8ue/stringh_file_not_found_cmake_on_osx_calls_wrong/), incorrect compiler may be called during this process.

**Comment by jschwe** [comment]
Interesting. Looking at the commits between f679545 and v0.1 I don't see anything obvious. Can you confirm that you are using the same CMake version as before?

Given the reddit thread, can you set `CMAKE_C_COMPILER` to `clang` when configuring cmake?  Also retest that including commit https://github.com/corrosion-rs/corrosion/commit/0b0b182f8d2e12a1558d12dc5aea15bc815c703b.

**Comment by niyue** [comment]
Thanks for the reply.

> Can you confirm that you are using the same CMake version as before?
Yes I did. Actually, simply switching between f679545a63a8b214a415e086f910126ab66714fa and v0.1 will show/avoid this error.

> can you set CMAKE_C_COMPILER to clang when configuring cmake? Also retest that including commit https://github.com/corrosion-rs/corrosion/commit/0b0b182f8d2e12a1558d12dc5aea15bc815c703b
My project setup is more complex than I said above, I used vcpkg in my project for managing dependencies, let me see how I can do some more testing here.

**Comment by niyue** [comment]
@jschwe  
I did some more testing, it is likely I made some mistake during previous verification and I said v0.1 didn't work above and this is incorrect, and I verified again and v0.1 (9e42e832d1ce207d6669a4ff898d705fdc3c98c2) works for me.

But the latest commit (904fefbedff0a781ea6db861100f584326adb19c) doesn't work with the same error mentioned above. And I tried verifying several older commits, and the commit fixing the json parsing (2fc942fbdeac179cb9f00cdde191c2d5b50614b7) reports the same issue, and I suspect this issue was introduced in commit  (0b0b182f8d2e12a1558d12dc5aea15bc815c703b) for setting `CXX` for cargo, but I am not able to verify this commit due to CMake parsing error (probably the json parsing error fixed later), with error message like below.

```
CMake Error at /Users/ss/dev/tools/vcpkg/installed/x64-osx/share/cmake/CorrosionGenerator.cmake:430 (string):
  string sub-command JSON failed parsing json string: * Line 1, Column 1

    Syntax error: value, object or array expected.

  .
Call Stack (most recent call first):
  /Users/ss/dev/tools/vcpkg/installed/x64-osx/share/cmake/Corrosion.cmake:349 (_generator_add_cargo_targets)
  cmake/rust_bridge.cmake:27 (corrosion_import_crate)
  cmake/rust_bridge.cmake:129 (add_library_rust)
  cmake/process_resources.cmake:1 (include)
  CMakeLists.txt:159 (include)
  CMakeLists.txt:290 (setup_src)


CMake Warning (dev) at /Users/ss/dev/tools/vcpkg/installed/x64-osx/share/cmake/CorrosionGenerator.cmake:431 (math):
  Unexpected character in expression at position 1: N

  Unexpected character in expression at position 2: O

  Unexpected character in expression at position 3: T

  Unexpected character in expression at position 4: F

  Unexpected character in expression at position 5: O

  Unexpected character in expression at position 6: U

  Unexpected character in expression at position 7: N

  Unexpected character in expression at position 8: D
```

**Comment by jschwe** [comment]
> I am not able to verify this commit due to CMake parsing error (probably the json parsing error fixed later), with error message like below.

Ah, so you are probably using a cmake version more recent than 3.19. Can you set the Cmake variable `CORROSION_EXPERIMENTAL_PARSER` to `0` (you mentioned this might be difficult in your setup). I enabled the experimental parser per default for 3.19 on master to get some feedback (to be sure it works before eventually enabling it for an actual release).

So just from reading your description, it sounds that your problem is the following:
-  we set `CC` and `CXX` to whatever cmake selects (or what is set in the environment)
- This causes problems for rust-krb5-src. Incidentally, what compiler is ` /Library/Developer/CommandLineTools/usr/bin/cc` on your system?

Can you set `CMAKE_C_COMPILER` or maybe the environment variables `CC`/`CXX` to whatever you need?

**Comment by niyue** [comment]
> so you are probably using a cmake version more recent than 3.19

Correct, I use the latest version of vcpkg, which uses cmake 3.22 internally

> Can you set the Cmake variable CORROSION_EXPERIMENTAL_PARSER to 0

I gave it a try, and it worked for me. I can verify that commit (0b0b182f8d2e12a1558d12dc5aea15bc815c703b) introduced this issue since its previous commit (b5ee2634ddd713af1bfdcd6fbf514646cb87ae30) can work.

> what compiler is /Library/Developer/CommandLineTools/usr/bin/cc on your system?

It is clang in Apple's command line tools. The env was listed in the first comment and it is like this:
```
macOS 12.3

❯ cc --version
Apple clang version 13.1.6 (clang-1316.0.21.2)
Target: x86_64-apple-darwin21.4.0
Thread model: posix
InstalledDir: /Library/Developer/CommandLineTools/usr/bin
```

> Can you set CMAKE_C_COMPILER or maybe the environment variables CC/CXX to whatever you need?

I am not certain what value I should change it to, since I just want to use Apple Clang, and it seems it is corrected detected. Let me see if I can figure it out.

**Comment by jschwe** [comment]
I have the suspicion, that the kerberos configure script mistakenly detects `cc` as a GCC compiler and then probably chooses some unsuitable flags. Can you compare the configure output with a successful build, and look if it also detects cc as gcc?

**Comment by tronical** [comment]
When using `/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/cc` (or `c++`) directly, it is necessary to provide the sysroot, in order for the compiler to know which version of macOS (or iOS, etc.) is being targeted. I guess the programs in `/usr/bin` will default to the default sdk sysroot.

In principle Corrosion could set `CFLAGS` and `CXXFLAGS` by repeating what CMake's generator does, using `CMAKE_<lang>_SYSROOT_FLAG` followed by `CMAKE_OSX_SYSROOT`. That I think would make builds work that use the `cc` crate. But I'm uncertain if it works with crates using the [cmake crate](https://crates.io/crates/cmake), because the invocation of cmake there will likely apply the sysroot twice: Once through corrosion provided `CXXFLAGS` and then through CMake's own logic.

**Comment by niyue** [comment]
Sorry I am not able to make further progress on this issue yet (many errands to handle due to the recent Shanghai lockdown), and will take a further look later.

**Comment by tronical** [comment]
I've thought about it a little and I think setting `SDKROOT` is a reasonable compromise. That fixes the issue for me locally - I propose #170.

**Comment by jschwe** [comment]
I think this issue should be resolved after #170. If this is not the case, feel free to reopen the issue.
